File: Models/Disp_V.cs
using ZC_ALM_TOOLS.Core;

namespace ZC_ALM_TOOLS.Models
{
    public class Disp_V : ObservableObject
    {
        // --- PROPIEDADES DEL EXCEL
        public string UID { get; set; }
        public int Numero { get; set; }
        public string Tag { get; set; }
        public string Descripcion { get; set; }
        public string FAT { get; set; }
        public string SByte { get; set; }
        public string SBit { get; set; }
        public string RRByte { get; set; }
        public string RRBit { get; set; }
        public string RTByte { get; set; }
        public string RTBit { get; set; }
        public string GrAlarma { get; set; }
        public string Cuadro { get; set; }
        public string Observaciones { get; set; }
        public string CPTag { get; set; }
        public string CPTipo { get; set; }
        public int CPNum { get; set; }
        public string CPComentario { get; set; }


        // --- PROPIEDAD PARA LA INTERFAZ ---
        // No se carga desde el CSV, se usa solo para mostrar resultados de comparación
        private string _Estado = "Sin comprobar";
        public string Estado
        {
            get => _Estado;
            set { _Estado = value; OnPropertyChanged(); }
        }

        // --- MÉTODO DE CARGA
        public static Disp_V FromCsv(string[] c) => new Disp_V
        {
            UID = DataHelper.GetVal(c, 0),
            Numero = DataHelper.ParseInt(DataHelper.GetVal(c, 1)),
            Tag = DataHelper.GetVal(c, 2),
            Descripcion = DataHelper.GetVal(c, 3),
            FAT = DataHelper.GetVal(c, 4),
            SByte = DataHelper.GetVal(c, 5),
            SBit =  DataHelper.GetVal(c, 6),
            RRByte = DataHelper.GetVal(c, 7),
            RRBit = DataHelper.GetVal(c, 8),
            RTByte = DataHelper.GetVal(c, 9),
            RTBit = DataHelper.GetVal(c, 10),
            GrAlarma = DataHelper.GetVal(c, 11),
            Cuadro = DataHelper.GetVal(c, 12),
            Observaciones = DataHelper.GetVal(c, 13),
            CPTag = DataHelper.GetVal(c, 14),
            CPTipo = DataHelper.GetVal(c, 15),
            CPNum = DataHelper.ParseInt(DataHelper.GetVal(c, 16)),
            CPComentario = DataHelper.GetVal(c, 17)
        };

    }


}

File: infrastructure/parsers/hardware/disp_v.py
"""
Hardware Parser - DispV
==========================
Extrae la tabla de dispositivos de Valvulas (V) desde el Excel Maestro.

30 columnas (vs 23 de ED/EA/SA) porque las valvulas tienen
campos especificos: S (setpoint), RR (retorno reposo), RT
(retorno trabajo) y 10 columnas Cfg.
"""

from core.models import DispV
from infrastructure.parsers.base_parser import BaseParser

__all__ = ["DispVParser"]


class DispVParser(BaseParser):
    """Parser para extraer la tabla de dispositivos de Valvulas."""

    def extraer(self, ruta_excel: str) -> list[DispV]:
        df = self._leer_tabla(
            ruta_excel=ruta_excel,
            sheet_name="DISP_V",
            table_name="Tabla_Disp_V",
            columnas_numericas=[
                # Columnas int del Excel
                "Numero",
                "S.Byte", "S.Bit",
                "RR.Byte", "RR.Bit",
                "RT.Byte", "RT.Bit",
                "Gr.Alarma",
                "PLC.Index",
                "Hmi.Index",
                # Las Cfg.* son strings (SCL crudo), NO entran aqui.
            ],
        )

        dispositivos: list[DispV] = []
        for _, row in df.iterrows():
            dispositivos.append(
                DispV(
                    # 4 atributos del Protocol
                    numero=int(row.get("Numero", 0)),
                    plc_tag=str(row.get("PLC.Tag", "")),
                    plc_comentario=str(row.get("PLC.Comentario", "")),
                    descripcion=str(row.get("Descripcion", "")),
                    # Excel extendido
                    uid=str(row.get("UID", "")),
                    tag=str(row.get("Tag", "")),
                    fat=str(row.get("FAT", "")),
                    s_byte=int(row.get("S.Byte", 0)),
                    s_bit=int(row.get("S.Bit", 0)),
                    rr_byte=int(row.get("RR.Byte", 0)),
                    rr_bit=int(row.get("RR.Bit", 0)),
                    rt_byte=int(row.get("RT.Byte", 0)),
                    rt_bit=int(row.get("RT.Bit", 0)),
                    gr_alarma=int(row.get("Gr.Alarma", 0)),
                    cuadro=str(row.get("Cuadro", "")),
                    observaciones=str(row.get("Observaciones", "")),
                    # PLC
                    plc_tipo=str(row.get("PLC.Tipo", "")),
                    plc_index=int(row.get("PLC.Index", 0)),
                    # HMI
                    hmi_index=int(row.get("Hmi.Index", 0)),
                    hmi_texto=str(row.get("Hmi.Texto", "")),
                    # Cfg (10 campos, todos son lineas SCL crudas)
                    cfg_habilitar=str(row.get("Cfg.Habilitar", "")),
                    cfg_byteretornoreposo=str(row.get("Cfg.ByteRetornoReposo", "")),
                    cfg_bitretornoreposo=str(row.get("Cfg.BitRetornoReposo", "")),
                    cfg_byteretornotrabajo=str(row.get("Cfg.ByteRetornoTrabajo", "")),
                    cfg_bitretornotrabajo=str(row.get("Cfg.BitRetornoTrabajo", "")),
                    cfg_byteactivacion=str(row.get("Cfg.ByteActivacion", "")),
                    cfg_bitactivacion=str(row.get("Cfg.BitActivacion", "")),
                    cfg_habitreposo=str(row.get("Cfg.HabRetReposo", "")),
                    cfg_habitrtrabajo=str(row.get("Cfg.HabRetTrabajo", "")),
                    cfg_grupoalarma=str(row.get("Cfg.GrupoAlarma", "")),
                )
            )
        return dispositivos
